liblangutil/SourceReferenceExtractor.cpp

```diff
@@ -106,7 +107,7 @@ SourceReference SourceReferenceExtractor::extract(SourceLocation const* _locatio
 		interest,
 		isMultiline,
 		line,
-		start.column,
-		end.column,
+		min(start.column, static_cast<int>(line.length())),
+		min(end.column, static_cast<int>(line.length()))
```

Comment:
This does prevent the crash but I'm not sure it actually fixes the underlying issue.

I think that the problem might actually be in `source->lineAtPosition()`. It seems that it strips a single  `\r` character from the line which is the reason why `start` ends up being past the end of the line. For example if the line is `\r\r\r\r`, it returns `\r\r\r`. This is inconsistent with ` source->translatePositionToLineColumn()`, which just ignores `\r` characters completely.

Also, even after the fix the locations shown in error messages seem off. For example for `\rabc` you get this:
```
Warning: Yul is still experimental. Please use the output with care.
Error: Expected keyword "object".
 --> /tmp/cr.yul:1:2:
  |
abc
  |  ^^^
```

Reply:
Yes, that is indeed the underlying issue, but I still think that this is the best fix. The problem with `\rabc` is a different one.

Reply:
Yeah, I think the problem is here:
https://github.com/ethereum/solidity/blob/661d11035b871aa1624bf568e3d476ef1bf25895/liblangutil/CharStream.cpp#L100-L101

It looks like it's meant to handle the `\r` left over after stripping `\n` on Windows. But for this to work correctly `translatePositionToLineColumn()` should also treat that `\r` in the same special way - as a part of the line ending - and it does not.

Reply:
`translatePositionToLineColumn()` could be causing similar problems in other places that we just haven't detected yet.

Reply:
I think `translatePositionToLineColumn` is perfectly implemented according to the specification. It is `lineAtPosition` that is questionable.

Reply:
We could change `lineAtPosition` to remove all trailing whitespace....

Reply:
Ok, no specification, but rather something like "the simplest definition of "line number" that would make sense".